### PR TITLE
[#IR-225] Add infinite scroll to sprint list

### PR DESCRIPTION
**Pull request by chetanyakan**

### CODE REVIEW

File: src/app/retrospective/sprint-list/sprint-list.component.ts [hunk 1]
```diff
@@ -22,14 +22,27 @@ export class SprintListComponent implements OnInit {
...
-        private router: Router,
-        private utils: UtilsService,
-        private changeDetectorRefs: ChangeDetectorRef
-    ) {
+    @HostListener('window:scroll', [])
+    onWindowScroll() {
+        const currentHeight = window.innerHeight + window.scrollY;
+
+        // If the user has scrolled to 70% of the page and is scrolling down, add more data
+        if (currentHeight > (document.body.offsetHeight * 0.7) && window.pageYOffset > this.lastScrollTop) {
```
Comment (himanshu-garg-165) [inline src/app/retrospective/sprint-list/sprint-list.component.ts]: Make `0.7` as a constant and also add some reference links (if any) for all the calculations so as to make the code more readable

File: src/app/retrospective/sprint-list/sprint-list.component.ts [hunk 2]
```diff
@@ -38,15 +51,26 @@ export class SprintListComponent implements OnInit {
...
         this.initializeDataSource();
-        this.changeDetectorRefs.detectChanges();
     }
 
     initializeDataSource() {
+        const documentBody = document.body, documentElement = document.documentElement;
+        // Taking the maximum value for the height of the page as it is
+        // calculated differently by different Browsers.
+        const pageHeight = Math.max( documentBody.scrollHeight, documentBody.offsetHeight,
+            documentElement.clientHeight, documentElement.scrollHeight, documentElement.offsetHeight );
```
Comment (himanshu-garg-165) [inline src/app/retrospective/sprint-list/sprint-list.component.ts]: Remove spaces around paranthesis

File: src/app/retrospective/sprint-list/sprint-list.component.ts [hunk 3]
```diff
@@ -38,15 +51,26 @@ export class SprintListComponent implements OnInit {
...
+
+        // Passing height/48 as rowsToLoad as min-height of
+        // mat-row is set to 48 by default.
```
Comment (himanshu-garg-165) [inline src/app/retrospective/sprint-list/sprint-list.component.ts]: Add some reference links

File: src/app/retrospective/sprint-list/sprint-list.component.ts [hunk 4]
```diff
@@ -38,15 +52,26 @@ export class SprintListComponent implements OnInit {
...
         this.dataSource = new SprintListDataSource(
             this.retrospectiveService,
             this.retrospectiveID,
+            Math.ceil(pageHeight / 48),
             this.showCannotGetSprintsError.bind(this)
         );
+
+        // Calling detectChanges here to refresh the value of isSprintsLoading.
+        this.changeDetectorRefs.detectChanges();
```
[the remark this answers is not in the document]
Comment (chetanyakan) [reply src/app/retrospective/sprint-list/sprint-list.component.ts]: As I have mentioned in the comment, we call detectChanges here to refresh the value of `isSprintsLoading`.

If we don't call `detectChanges` here, we get ``` “expression has changed after it was checked” exception```.

File: src/app/retrospective/sprint-list/sprint-list.component.ts (hunk 3 above)
Comment (chetanyakan) [reply src/app/retrospective/sprint-list/sprint-list.component.ts]: You can see [here](https://github.com/angular/material2/blob/2046b7a3287bbff87a48800cf7d770f1976ed757/src/lib/table/table.scss#L2) that $mat-row-height is set to 48px.

You can see [here](https://github.com/angular/material2/blob/2046b7a3287bbff87a48800cf7d770f1976ed757/src/lib/table/table.scss#L17) that min-height of mat-row is set to `$mat-row-height`.

File: src/app/retrospective/sprint-list/sprint-list.component.ts [hunk 5]
```diff
@@ -16,12 +16,29 @@ import { UtilsService } from '../../shared/utils/utils.service';
...
+
+    @Input() retrospectiveID: any;
+
+    @HostListener('window:scroll', [])
+    onWindowScroll() {
+        const currentHeight = window.innerHeight + window.scrollY;
+        const minScrollHeight = document.body.offsetHeight * LOAD_NEXT_PAGE_SCROLL_POSITION;
+
+        // If the user has scrolled to 70% of the page and is scrolling down, add more data
+        if (currentHeight > minScrollHeight && window.pageYOffset > this.lastScrollTop) {
```
Comment (anmolghosh) [inline src/app/retrospective/sprint-list/sprint-list.component.ts]: check if we have any good packages to support infinite scrolling. Also, if not found, we should develop this as a re-usable service/components

File: src/app/retrospective/sprint-list/sprint-list.data-source.ts [hunk 6]
```diff
@@ -39,9 +43,6 @@ export class SprintListDataSource extends DataSource<any> {
                 },
                 err => {
                     this.errorCallback(err);
-                },
-                () => {
```
Comment (himanshu-garg-165) [inline src/app/retrospective/sprint-list/sprint-list.data-source.ts]: The complete callback is not fired if the error callback is fired (Refer https://github.com/angular/angular/issues/7865)

File: src/app/retrospective/sprint-list/sprint-list.component.ts (hunk 5 above)
Comment (himanshu-garg-165) [reply src/app/retrospective/sprint-list/sprint-list.component.ts]: @anmolghosh Added a TODO for the same and creating a ticket also

**Pull request merged**